File: backend_v2/app/schemas/employee.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class EmployeeCreate(BaseModel):
    first_name: str = Field(min_length=1, max_length=80)
    last_name: str = Field(min_length=1, max_length=120)
    email: str | None = Field(default=None, max_length=255)
    phone: str | None = Field(default=None, max_length=40)
    dni: str | None = Field(default=None, max_length=40)
    role_title: str | None = Field(default=None, max_length=100)
    pin: str | None = Field(default=None, min_length=4, max_length=4)
    password: str | None = Field(default=None, min_length=8, max_length=256)
    hired_on: date | None = None
    is_active: bool = True

    @field_validator("first_name", "last_name", "email", "phone", "dni", "role_title", "pin", mode="before")
    @classmethod
    def strip_optional_strings(cls, value: object) -> object:
        if isinstance(value, str):
            stripped = value.strip()
            return stripped or None
        return value

    @field_validator("email", mode="after")
    @classmethod
    def normalize_email(cls, value: str | None) -> str | None:
        return value.lower() if value else value

    @field_validator("pin")
    @classmethod
    def validate_pin(cls, value: str | None) -> str | None:
        if value and not value.isdigit():
            raise ValueError("PIN must contain digits only.")
        return value

    @model_validator(mode="after")
    def password_requires_email(self) -> "EmployeeCreate":
        if self.password and not self.email:
            raise ValueError("An email address is required when setting a password.")
        return self


class EmployeeUpdate(BaseModel):
    first_name: str | None = Field(default=None, min_length=1, max_length=80)
    last_name: str | None = Field(default=None, min_length=1, max_length=120)
    email: str | None = Field(default=None, max_length=255)
    phone: str | None = Field(default=None, max_length=40)
    dni: str | None = Field(default=None, max_length=40)
    role_title: str | None = Field(default=None, max_length=100)
    is_active: bool | None = None

    @field_validator("first_name", "last_name", "email", "phone", "dni", "role_title", mode="before")
    @classmethod
    def strip_strings(cls, value: object) -> object:
        if isinstance(value, str):
            stripped = value.strip()
            return stripped or None
        return value


class EmployeeRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    company_id: UUID
    user_id: UUID | None
    first_name: str
    last_name: str
    full_name: str
    email: str | None
    phone: str | None
    dni: str | None
    role_title: str | None
    hired_on: date | None
    is_active: bool
    has_pin: bool
    created_at: datetime
    updated_at: datetime


class EmployeeListResponse(BaseModel):
    items: list[EmployeeRead]
    total: int
    limit: int | None
    offset: int

File: frontend-next/types/employee.ts
export type EmployeeRole = "owner" | "admin" | "hr_manager" | "manager" | "employee";

export interface Employee {
  id: string;
  company_id: string;
  user_id: string | null;
  first_name: string;
  last_name: string;
  full_name: string;
  initials: string;
  email: string | null;
  phone: string | null;
  dni: string | null;
  role_title: string | null;
  hired_on: string | null;
  is_active: boolean;
  has_pin: boolean;
  schedule_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface EmployeeCreateRequest {
  first_name: string;
  last_name: string;
  dni?: string;
  password?: string;
  pin?: string;
  email?: string;
  phone?: string;
  role_title?: string;
  hired_on?: string;
  is_active?: boolean;
}

export interface CsvRowError {
  row: number;
  field: string;
  message: string;
}

export interface CsvImportResult {
  imported: number;
  skipped: number;
  errors: CsvRowError[];
  preview: Record<string, string>[];
}

export interface EmployeeUpdateRequest {
  first_name?: string;
  last_name?: string;
  dni?: string;
  email?: string;
  phone?: string;
  role_title?: string;
  hired_on?: string;
  password?: string;
  pin?: string | null;
  is_active?: boolean;
  schedule_id?: string | null;
}
